File: fence.cpp
/*
ID: totheso1
LANG: C++14
TASK: fence
*/

#include <algorithm>
#include <forward_list>
#include <fstream>
#include <iterator>
#include <stack>
#include <stdexcept>
#include <unordered_map>
#include <utility>
#include <vector>

using namespace std;

static ifstream fin("fence.in");
static ofstream fout("fence.out");

#define endl '\n'

template <typename T>
T fin_get() {
  return *istream_iterator<T>(fin);
}

template <typename C>
C fin_get_collection(const int size) {
  C ret;
  copy_n(istream_iterator<typename C::value_type>(fin), size,
         back_inserter(ret));
  return ret;
}

static const int FENCE_COUNT = fin_get<int>();
static unordered_map<int, int> degree_by_point_collection;
static unordered_map<int, forward_list<pair<int, int>>>
    edge_collection_by_point_collection;

static void initialize() {
  const auto add_edge = [](const int f, const int t, const int idx) {
    edge_collection_by_point_collection[f].emplace_front(t, idx);
  };

  for (int edge_id = 0; edge_id < FENCE_COUNT; ++edge_id) {
    const int point_a = fin_get<int>();
    const int point_b = fin_get<int>();
    ++degree_by_point_collection[point_a];
    ++degree_by_point_collection[point_b];
    add_edge(point_a, point_b, edge_id);
    add_edge(point_b, point_a, edge_id);
  }
}

static void solve() {
  vector<bool> traveled(FENCE_COUNT, false);
  vector<int> sequence_reversed;

  const auto recursive_find_path = [&](const int curr_point) {
    const auto impl = [&](const auto &self, const int curr_point) -> void {
      for (const auto &p : edge_collection_by_point_collection[curr_point]) {
        const int to = p.first;
        const int edge_id = p.second;
        if (!traveled[edge_id]) {
          traveled[edge_id] = true;
          self(self, to);
        }
      }
      sequence_reversed.push_back(curr_point);
    };
    return impl(impl, curr_point);
  };

  const auto non_recursive_find_path = [&](const int start_point) {
    auto b = cbegin(edge_collection_by_point_collection[start_point]);
    const auto e = cend(edge_collection_by_point_collection[start_point]);
    using iter_type = decltype(b);
    stack<pair<int, pair<iter_type, const iter_type>>> stk;

    const auto in_the_loop = [&](const int curr_point, auto &b, const auto &e) {
      for (; b != e; ++b) {
        const int to = b->first;
        const int edge_id = b->second;
        if (!traveled[edge_id]) {
          traveled[edge_id] = true;
          stk.emplace(to,
                      make_pair(cbegin(edge_collection_by_point_collection[to]),
                                cend(edge_collection_by_point_collection[to])));
          return;
        }
      }
      sequence_reversed.push_back(curr_point);
      stk.pop();
    };

    stk.emplace(start_point, make_pair(b, e));
    while (!stk.empty()) {
      in_the_loop(stk.top().first, stk.top().second.first,
                  stk.top().second.second);
    }
  };

  vector<int> ending_point_collection;
  for (const auto &p : degree_by_point_collection) {
    const int point = p.first;
    const int degree = p.second;
    if (degree & 1) {
      ending_point_collection.push_back(point);
    }
  }
  int start;
  if (!ending_point_collection.empty()) {
    if (ending_point_collection.size() != 2) {
      throw invalid_argument("No possible Eulerian circuit.");
    }
    start =
        min(ending_point_collection.front(), ending_point_collection.back());
  } else {
    start = min_element(cbegin(degree_by_point_collection),
                        cend(degree_by_point_collection))
                ->first;
  }

  for (auto &p : edge_collection_by_point_collection) {
    p.second.sort();
  }
  non_recursive_find_path(start);
  copy(crbegin(sequence_reversed), crend(sequence_reversed),
       ostream_iterator<int>(fout, "\n"));
}

int main() {
  initialize();
  solve();
}

File: fence.py
"""
ID: totheso1
LANG: PYTHON3
TASK: fence
NOTE:
    Python3 will get
    "RecursionError:_maximum_recursion_depth_exceeded_in_comparison"
    if use a naive recursive depth-first-search.
    Rewrite a non-recursive version or use C++/Java/Pascal instead.
"""

from itertools import islice
from typing import Any, Counter, DefaultDict, List, Tuple

fin = open("fence.in", "r")
fout = open("fence.out", "w")


def fprint(*args: Any, **kwargs: Any) -> None:
    print(*args, file=fout, **kwargs)


# Code start
FENCE_COUNT = int(fin.readline())
degree_by_point_collection = Counter[int]()
edge_collection_by_point_collection = DefaultDict[int, List[Tuple[int, int]]](list)


def initialize() -> None:
    def add_edge(f: int, t: int, idx: int) -> None:
        edge_collection_by_point_collection[f].append((t, idx))

    for edge_id, line in enumerate(fin):
        point_a, point_b = map(int, line.split())
        degree_by_point_collection[point_a] += 1
        degree_by_point_collection[point_b] += 1
        add_edge(point_a, point_b, edge_id)
        add_edge(point_b, point_a, edge_id)


def solve() -> None:
    def recursive_find_path(curr_point: int) -> None:
        for to, edge_id in edge_collection_by_point_collection[curr_point]:
            if not traveled[edge_id]:
                traveled[edge_id] = True
                recursive_find_path(to)
        sequence_reversed.append(curr_point)

    def non_recursive_find_path(start_point: int) -> None:
        def in_the_loop(curr_point: int, i: int) -> None:
            for to, edge_id in islice(
                edge_collection_by_point_collection[curr_point], i, None
            ):
                stack[-1][1] += 1
                if not traveled[edge_id]:
                    traveled[edge_id] = True
                    stack.append([to, 0])
                    return
            sequence_reversed.append(curr_point)
            stack.pop()

        stack = [[start_point, 0]]
        while stack:
            in_the_loop(*stack[-1])

    ending_point_collection = []
    for point, degree in degree_by_point_collection.items():
        if degree & 1:
            ending_point_collection.append(point)
    if ending_point_collection:
        if len(ending_point_collection) != 2:
            raise ValueError("No possible Eulerian circuit.")
        start = min(ending_point_collection)
    else:
        start = min(degree_by_point_collection)

    for edge_collection in edge_collection_by_point_collection.values():
        edge_collection.sort()
    sequence_reversed: List[int] = []
    traveled = [False] * FENCE_COUNT
    non_recursive_find_path(start)
    fprint(*reversed(sequence_reversed), sep="\n")


initialize()
solve()
# Code end
